doc/design/scheduled-tasks.md

```diff
@@ -34,21 +34,46 @@ Whenever a Version is modified, a flag `needs_recomputing` is set on the Version
 
 #### Scheduled job
 Separately, a scheduled job that runs on a configurable fairly tight interval will check for any Versions that need to be updated.
-For each of those Versions, it also checks that the Version hasn't been modified in a certain amount of time (say 5 minutes). This prevents DB locks from happening while Assets are still being uploaded to the Version.
 
-The code would look something like this:
+The full code would look something like this:
 ```
 def scheduled_task():
   versions = Version.objects.filter(needs_recomputing=True).filter(modified__lt=timezone.now() - 5 minutes)
   # SELECT ... FROM api_version WHERE api_version.needs_recomputing AND now() - api_version.modified > 5 minutes;
+
+  # Mark all versions as not needing recomputation to prevent subsequent tasks from picking them up as well.
+  for version in versions:
+    version.needs_recomputing = False
+    version.save()
+
+  for version in versions:
+    try:
+      # Calculate assetsSummary
+      version.metadata['assetsSummary'] = calculate_assetsSummary(version)
+
+      # Validate the version
+      validate_version(version)
```

Comment:
since we are not locking DB (`version`) from modification for the duration of this task, above code might be racy with possible users' modifications. Some race could be avoided if this `validate_version` would not set validation status `if version.needs_recomputing` -- i.e. that something was updated/uploaded before it completed.  Cons: leaks "needs_recomputing" into "validate_version" which should in principle not care about all this orchestration, so may be it should just return the value either it is valid or not and logic should stay here to assign (like for `assetsSummary`)

Reply:
Good point, I changed the design to lock the DB for the duration of the task. In practice multiple Versions will very rarely be picked up at once, since a single user is generally only with a single version at a time.
If we ever run into excessive DB locks in the future, we can scale this behavior back.